### ISSUE TITLE
Incorrect link preview for non-UTF-8 sites

### ISSUE THREAD

**Issue by mike83cz** [humhub/humhub-modules-linkpreview#16]
For example, if I want to share a page encoded in "Windows-1250", the title and description of the preview will end with the first Latin extended character.

Can this somehow be corrected/converted?

![HumHub-linkpreview-issue](https://user-images.githubusercontent.com/87585815/126901327-c7038e5a-8ae1-4a5d-a130-c78b5fab2037.png)

Title should be "ANKETA: Jak vnímáte poslední kroky Ubisoftu?"

**Comment by yurabakhtin** [comment]
@luke- Please see what response:

![response_chars](https://user-images.githubusercontent.com/6995524/126954231-b011bbe9-bc05-4d69-9ec4-120ea8621495.jpg)

We do the request by code `client.post(this.fetchUrl, {data: {url: this.url}})`, but even if I use `client.post(this.fetchUrl, {data: {url: this.url}, headers: {'Content-Type': 'application/x-www-form-urlencoded; charset=windows-1250'}})` I didn't get any effect, but we cannot use this way because we cannot know charset of the requested page.

I think this problem only of the page and all data should encoded to html entities on the page side like it is partly done for the meta tag `twitter:description` there.

**Comment by luke-** [comment]
@yurabakhtin Thanks for testing!
@mike83cz Can you please provide other example sites?

**Comment by mike83cz** [comment]
Yes, for example another huge Czech news portal lidovky.cz, also with "Windows-1250" charset. Same result.

![lidovky](https://user-images.githubusercontent.com/87585815/127063291-88835a8c-8b49-4a9c-b48c-76f9b941ffe0.png)

link https://www.lidovky.cz/svet/video-v-norsku-spadl-velky-meteorit-jasne-rozzaril-nocni-oblohu.A210725_172704_ln_zahranici_lijk

Since f.e. on Facebook and other major networks the thumbnails from both sites are displayed correctly, I was hoping that this could be solved somehow.

I would say the problem is the charset, but I'm not a developer, I'd be grateful for any advice.

![Facebook](https://user-images.githubusercontent.com/87585815/127062664-96b5d587-aa4f-41ca-b78f-7d623cbac23e.png)

**Comment by mike83cz** [comment]
Next example Hospodarske Noviny (hn.cz) - charset Windows-1250

![hn-lp](https://user-images.githubusercontent.com/87585815/129440394-df077563-660b-4889-97b3-31e7b4f92324.png)

link: https://archiv.hn.cz/c7-66962530-rt35g-df15f8ad901f593

Same issue. All these webs (idnes.cz, lidovky.cz, hn.cz) are the largest (top 10) news portals in my country, people share it very often.

**Comment by yurabakhtin** [comment]
@luke- @mike83cz Fixed in PR https://github.com/humhub/linkpreview/pull/4:

<img width="1826" alt="fixed_link_preview" src="https://user-images.githubusercontent.com/6995524/129558307-f3116f43-ef55-4663-8518-f536faabcf94.png">

**Comment by luke-** [comment]
@yurabakhtin Thanks!

**Comment by mike83cz** [comment]
Hey guys, it's me again.
I excitedly updated (from 1.15.3) to the latest version 1.15.6 with the prospect that the problems with the Czech language are solved, but unfortunately the situation is much worse than before.

Before, only a few Czech websites that didn't use UTF-8 had problems, but now the Czech language is displayed incorrectly on all websites, including YouTube, etc... With the exception of hn.cz, which is the only one that works well.

The major problem is the UTF-8 sites are broken now
Novinky.cz (UTF-8)
![novinky](https://user-images.githubusercontent.com/87585815/140614042-b2d78e32-ec1c-4391-9aa6-3ff01f99e873.png)

link: https://www.novinky.cz/kultura/filmy-serialy/clanek/marvel-predstavuje-novou-partu-do-kin-vtrhli-eternals-40376946

Youtube
![Videohry-Gamehold](https://user-images.githubusercontent.com/87585815/140614075-6bc0fee0-f259-46e9-a0f2-a83cfead6705.png)

link https://youtu.be/EPXYWDJbEfA

Not UTF-8 sites got better, but still not good
Idnes.cz (Windows-1250)
![idnes](https://user-images.githubusercontent.com/87585815/140614399-ce3ff282-1e26-4908-aaf3-5c283a3c0b85.png)

link https://www.idnes.cz/hry/recenze/new-world-recenze-amazon.A211101_143008_bw-pc-recenze_oz

lidovky.cz
![lidovky](https://user-images.githubusercontent.com/87585815/140614535-498d6d31-6035-44e3-a510-699d883dd6b7.png)

link https://www.lidovky.cz/svet/video-v-norsku-spadl-velky-meteorit-jasne-rozzaril-nocni-oblohu.A210725_172704_ln_zahranici_lijk

Is there an option of back roll? I need solve this ASAP. Or I would try go back before issue #18 fix.

Thanks, Mike

**Comment by luke-** [comment]
@mike83cz please contact our support at hello@humhub.com to request an older version.

**Comment by yurabakhtin** [comment]
@luke- Unfortunately I cannot find another better solution.

In the method `humhub\modules\linkpreview\controllers\IndexController::fetch($url)` we get a page content by CURL request. I tried to detect an encoding there and then convert to UTF-8 but I see same result.
My test code there:

```php
$encoding = mb_detect_encoding($body, mb_list_encodings());
$body = utf8_decode(mb_convert_encoding($body, mb_internal_encoding(), $encoding));

// $encoding == 'ISO-8859-1'
// mb_internal_encoding() == 'UTF-8'
```

---
Test for `https://www.idnes.cz/hry/recenze/new-world-recenze-amazon.A211101_143008_bw-pc-recenze_oz`:

Without my code I get:

```
Ambiciózní online hra New World zaznamenala u hráèù úspìch, zároveò ji ovem trápí øada porodních bolestí. Nyní bude záleet, jak rychle a v jaké kvalitì dokáe Amazon dodat dalí obsah.
```

If I don't apply `utf8_decode()`:

```
AmbiciÃ³znÃ­ online hra New World zaznamenala u hrÃ¡Ã¨Ã¹ ÃºspÃ¬ch, zÃ¡roveÃ² ji ovÂem trÃ¡pÃ­ Ã¸ada porodnÃ­ch bolestÃ­. NynÃ­ bude zÃ¡leÂet, jak rychle a v jakÃ© kvalitÃ¬ dokÃ¡Âe Amazon dodat dalÂÃ­ obsah.
```

With my code I get the same result:

```
Ambiciózní online hra New World zaznamenala u hráèù úspìch, zároveò ji ovem trápí øada porodních bolestí. Nyní bude záleet, jak rychle a v jaké kvalitì dokáe Amazon dodat dalí obsah.
```

It looks like the CURL request already returns the broken chars, but I don't find a solution to get a correct result even if I call the CURL with headers `->addHeaders(['Content-Type' => 'text/html; charset=UTF-8'])`.

**Comment by mike83cz** [comment]
Vesion 1.15.5 makes no difference. Behavior is the same as 1.15.6
Back roll to 1.15.3 returned the site to its original functionality. Czech UTF-8 sites works well.

**Comment by luke-** [comment]
@Soundwave2142 Would you like to take a look at the issue?

**Comment by luke-** [comment]
https://github.com/humhub/linkpreview-issues/issues/18#issuecomment-980340247

**Comment by aaukt** [comment]
I found multiple issues with this.

First, mb_detect_encoding() should not be the first way to get the encoding. As per https://www.php.net/manual/de/function.mb-detect-encoding.php

> It is always preferable to use an indication of character encoding stored or transmitted with the data, such as a "Content-Type" HTTP header.

For most of the czech links the returned encoding is actually windows-1250 (it is almost ISO-8859-1 but not quite).

Second, for reliable parsing in DOMDocument i found it is best to convert UTF-8 to HTML-ENTITIES.

So with the following code i could get the correct results for all links posted and also other ones that didn't work before for me (german umlauts).

```
$contentType = $response->getHeaders()->get('Content-Type', '');
if (preg_match('/charset=(.*)/i', $contentType, $matches)) {
    $encoding = strtoupper($matches[1]);
} elseif (($encoding = mb_detect_encoding($body)) !== false) {
    $encoding = strtoupper($encoding);
} else {
    $encoding = 'UTF-8';
}

if (in_array($encoding, array_map('strtoupper', mb_list_encodings()))) {
    $body = mb_convert_encoding($body, mb_internal_encoding(), $encoding);
} else {
    $body = iconv($encoding, 'UTF-8', $body);
}

$output = mb_convert_encoding($body, 'HTML-ENTITIES', 'UTF-8');
```

**Comment by matll42** [comment]
Hi, we have the same with French accents encoding but not with all websites

Here is the news from Radio Canada with charset UTF-8 : https://ici.radio-canada.ca/nouvelle/1854041/quebec-legault-covid-rentree-couvre-feu

So we see that 
![image](https://user-images.githubusercontent.com/1232923/149410858-131379b4-8a7a-4892-8953-5dc74dabad7c.png)

This one is working lapresse with also charset UTF-8 : https://www.lapresse.ca/covid-19/2022-01-13/le-couvre-feu-leve-lundi/un-retour-en-classe-securitaire-assure-quebec.php
and we see that
![image](https://user-images.githubusercontent.com/1232923/149411143-ae89cbe9-f941-41c2-97a5-7f85fe2743fc.png)

I'm using version 1.14 with Humhub 1.7.1. We will try to update humhub but we customized the core so not so simple.

**Comment by matll42** [comment]
follow up : we also tried an upgrade LinkPreview to 1.15.6 with Humhub 1.9.2

**Comment by luke-** [comment]
Can you please try it again with the latest version 1.16.1? Please note PHP 7.4 is the minimum requirement.

**Comment by mike83cz** [comment]
I can't see this update. Last compatible version I can update on is 1.15.6. What is required Humhub version? I run 1.8.0. (with PHP 7.4.15)...

**Comment by luke-** [comment]
HumHub 1.9+ is required for the refactored link preview version.

**Comment by mike83cz** [comment]
Hey guys,

Please accept my apologies for the time delay, but unfortunately I didn't really have time to do a big Humhub update (many customizations) earlier.

I am now running Humhub 1.13 with LivePreview plugin 1.16.1 and unfortunately the situation is still not resolved.

Fortunately UTF-8 sites run perfectly fine, but Windows-1250 sites show corrupted characters in the preview. However, improvement has come at least in that the previews are displayed complete and don't end with the first corrupted character.

Some examples here:
idnes.cz - https://www.idnes.cz/hry/magazin/komentar-zaklinac-witcher-3-next-gen-verze-dlc-cd-projekt.A230104_114456_bw-magazin_oma
![idnes](https://user-images.githubusercontent.com/87585815/210900864-0e56a6d8-5bad-46d9-8b63-42013db8c687.png)

hn.cz - https://tech.hn.cz/c1-67155230-jak-se-vyrovnat-s-depresivnim-vyhledem-na-rok-2023-ve-svete-technologii-staci-necekat-zazraky
![hn](https://user-images.githubusercontent.com/87585815/210901021-09e1c6a2-fa8e-42be-a379-863cf45e56f7.png)

lidovky.cz - https://www.lidovky.cz/orientace/kultura/kultura-film-serial-indiana-jones-masinove-2023-rok-vyhled.A221229_164106_ln_kultura_rkj
![lidovky](https://user-images.githubusercontent.com/87585815/210901268-e3fc11fb-4d21-4df2-b301-69753b901bc4.png)

Thanks so much for your efforts guys, I believe we'll figure it out eventually.

**Comment by gevorgmansuryan** [comment]
Should be fixed in https://github.com/humhub/linkpreview/pull/11

**Comment by luke-** [comment]
Released 2.0.0-beta.1

**Comment by mike83cz** [comment]
I tried 2.0.0-beta.1 on a test instance and it looks like it's actually resolved. Thanks a lot. I'm looking forward to a stable 2.0 release.
